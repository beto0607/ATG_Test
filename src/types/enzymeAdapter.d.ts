// To avoid a typescript error with enzyme-adaoter-react-16
declare module 'enzyme-adapter-react-16';