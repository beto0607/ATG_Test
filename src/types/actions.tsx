import { GameSchedule, GameData } from "../types";

export const GAME_SCHEDULE_LOADED = "GAME_SCHEDULE_LOADED";
export type GAME_SCHEDULE_LOADED = typeof GAME_SCHEDULE_LOADED;

export const GAME_DATA_LOADED = "GAME_DATA_LOADED";
export type GAME_DATA_LOADED = typeof GAME_DATA_LOADED;


export interface GameScheduleLoadedAction {
    type: typeof GAME_SCHEDULE_LOADED;
    text: string;
    gameSchedule?: GameScheduleÂ | null;
}

export interface GameDataLoadedAction {
    type: typeof GAME_DATA_LOADED;
    gameData?: GameData;
}

export type LoadedActions = GameScheduleLoadedAction | GameDataLoadedAction;

export type AppActions = LoadedActions;