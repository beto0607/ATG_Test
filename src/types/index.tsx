type TrackInfo = {
    id: number,
    name: string
};
type FavoriteInfo = {
    raceId: string,
    legNumber: number,
    start: {
        number: number,
        name: string,
        distribution: number
    }
};
type GameInfo = {
    id: string,
    startTime: string, // YYYY-MM-DDThh:mm:ss
    tracks: Array<TrackInfo>,
    favorites: Array<FavoriteInfo>
}
//ENDPOINT RESPONSE
//https://www.atg.se/services/racinginfo/v1/api/products/<gameType>
type GameSchedule = {
    betType: string,
    upcoming: Array<GameInfo>,
    results: Array<GameInfo>
};

type RaceStart = {
    number: number,
    driver: {
        firstName: string,
        lastName: string
    },
    horse: {
        name: string,
        trainer: {
            firstName: string,
            lastName: string
        },
        pedigree: {
            father: {
                name: string
            }
        }
    }
}
type GameRace = {
    date: string,
    name: string,
    scheduledStartTime: string,
    starts: Array<RaceStart>
}
//ENDPOINT RESPONSE
// https://www.atg.se/services/racinginfo/v1/api/games/<gameId>
type GameResponse = {
    id: string,
    status: string,
    races: Array<GameRace>
};

export interface ApplicationState {
    text: string;
    game_schedule: GameSchedule;
    game_data: GameResponse;
}