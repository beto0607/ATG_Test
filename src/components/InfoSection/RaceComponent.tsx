import React, { useState } from "react";
import { Collapse, Badge, Alert } from "react-bootstrap";
import { Race, RaceStart } from '../../types/index';
import Moment from "react-moment";
import styles from './race.module.scss';

export const StartComponent: React.FC<RaceStart> = ({ number, horse, driver }: RaceStart) => {
    const [open, setOpen] = useState(false);
    return (
        <div onClick={() => setOpen(!open)} className={styles['start-container']}>
            <div className={styles['start-title-container']}>
                <Badge variant="info">Start: #{number}</Badge>
                <span>Horse: <strong>{horse.name}</strong></span>
                <span>Driver: <strong>{`${driver.firstName} ${driver.lastName}`}</strong></span>
            </div>
            <Collapse in={open}>
                <div>
                    <span>Trainer: {`${horse.trainer.firstName} ${horse.trainer.lastName}`}</span>
                    <span>Horse father: {`${horse.pedigree.father.name}`}</span>
                </div>
            </Collapse>
        </div>
    )
};
export const RaceComponent: React.FC<Race> = ({ id, name, number, scheduledStartTime, starts }: Race) => {
    return (
        <div className={styles['race-container']}>
            <div className={styles['race-title-container']}>
                <Alert variant='info'>
                    <span>#{number} {name && <strong>- {name}</strong>}</span>
                </Alert>
                <Alert variant='info'>
                    Start time:
                    <Moment date={scheduledStartTime} format='HH:mm:ss' />
                </Alert>
            </div>
            {starts.map(start => <StartComponent key={`${id}_${start.number}`} {...start} />)}
        </div>
    );
};