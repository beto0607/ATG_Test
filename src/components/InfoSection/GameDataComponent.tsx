import React from "react";
import { connect } from "react-redux";
import { loadGameData } from "../../actions/actions";
import { ApplicationState, GameData, Race } from '../../types/index';
import { RaceComponent } from './RaceComponent';

export interface GameDataProps {
    parent_id: string;
    gameData: GameData | null;
    loadGameData: any;
    load: boolean;
}
export const GameDataComponentConnected: React.FC<GameDataProps> = ({ parent_id, load, gameData, loadGameData }: GameDataProps) => {
    // Loads the game data only when the component is visible and only one time.
    if (load && !gameData) { loadGameData(parent_id); }
    return (
        <div>
            {gameData !== null && gameData.races.map((race: Race) => <RaceComponent key={race.id} {...race} />)}
        </div>
    );
};
const mapStateToProps = (state: ApplicationState, ownprops: any): GameData | null => {
    return {
        ...ownprops,
        gameData: state.gameData.find((e: GameData) => e.id === ownprops.parent_id) || null
    }
};
const mapDispatchToPropsGameData = (dispatch: any, ownprops: any): any => {
    return {
        ...ownprops,
        loadGameData: (gameId: string) => dispatch(loadGameData(gameId))
    };
}
export const GameDataComponent = connect(mapStateToProps, mapDispatchToPropsGameData)(GameDataComponentConnected);
