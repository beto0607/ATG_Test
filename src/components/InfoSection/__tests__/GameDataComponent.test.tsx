import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json'
import { GameData } from '../../../types/index';
import { GameDataProps, GameDataComponentConnected } from '../GameDataComponent';
import { RaceComponent } from '../RaceComponent';

const mockGameData: GameData = {
    id: 'id',
    status: 'status',
    races: [{
        id: 'race_1',
        number: 1,
        scheduledStartTime: '',
        starts: []
    }]
};
const mockGameDataProps: GameDataProps = {
    parent_id: '',
    gameData: null,
    loadGameData: jest.fn(),
    load: false
};


describe('GameDataComponentConnected', () => {
    it('renders without crashing', () => {
        const props: GameDataProps = mockGameDataProps;
        const component = shallow(<GameDataComponentConnected {...props} />);
        expect(toJson(component)).toMatchSnapshot()
    });
    it('should call loadGameData when load is true', () => {
        const props: GameDataProps = {
            ...mockGameDataProps,
            load: true,
            loadGameData: jest.fn()
        };
        shallow(<GameDataComponentConnected {...props} />);
        expect(props.loadGameData).toHaveBeenCalled()
    });
    it('should call loadGameData when load is true', () => {
        const props: GameDataProps = {
            ...mockGameDataProps,
            load: true,
            loadGameData: jest.fn()
        };
        shallow(<GameDataComponentConnected {...props} />);
        expect(props.loadGameData).toHaveBeenCalled()
    });
    it('should render 1 RaceComponent', () => {
        const props: GameDataProps = {
            ...mockGameDataProps,
            gameData: mockGameData
        };
        const component = shallow(<GameDataComponentConnected {...props} />);
        expect(component.find(RaceComponent)).toHaveLength(1);
    })
});
