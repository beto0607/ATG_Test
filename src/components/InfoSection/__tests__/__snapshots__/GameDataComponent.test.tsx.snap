// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameDataComponentConnected renders without crashing 1`] = `<div />`;
